File: internal/agent/protocol.go
// Package agent implements the privileged side of OPanel.
//
// opanel-api runs unprivileged and can do nothing to the host on its own. It
// asks opanel-agent, which runs as root behind a unix socket, to carry out
// each privileged step. Every action is a typed Go function with a validated
// input struct, and every external command is executed with an argv slice --
// never a shell string. That is the whole point of the split: it removes
// shell quoting, word splitting and injection from the threat model instead
// of trying to filter them out with regular expressions.
package agent

import "encoding/json"

// SocketPath is the default unix socket. Mode 0660, owned root:opanel.
const SocketPath = "/run/opanel/agent.sock"

// MaxMessageBytes caps a single request or response. Actions move config
// text, not files; anything larger is a bug or an attack.
const MaxMessageBytes = 4 << 20 // 4 MiB

// Request is one call. The wire format is a single JSON object followed by a
// newline; the connection carries exactly one request and one response.
type Request struct {
	// ID is echoed back so a caller can correlate logs. It has no protocol
	// meaning: one connection carries one exchange.
	ID string `json:"id,omitempty"`
	// Action names the handler, e.g. "ping" or "systemd.restart".
	Action string `json:"action"`
	// Version pins the payload shape. An action bumps it when its input
	// struct changes incompatibly, so a stale api binary fails loudly
	// instead of being misread by a newer agent.
	Version int `json:"version"`
	// Payload is the action-specific input, decoded by the handler.
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Response is the single reply to a Request.
type Response struct {
	ID      string          `json:"id,omitempty"`
	OK      bool            `json:"ok"`
	Error   string          `json:"error,omitempty"`
	Code    string          `json:"code,omitempty"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Error codes returned in Response.Code so callers can branch without
// string-matching the message.
const (
	CodeUnknownAction   = "unknown_action"
	CodeVersionMismatch = "version_mismatch"
	CodeBadPayload      = "bad_payload"
	CodeDenied          = "denied"
	CodeInternal        = "internal"
	CodeTimeout         = "timeout"
)

File: crates/opanel-agent/src/protocol.rs
//! The wire format between the panel and the agent.
//!
//! One JSON object, a newline, one reply, and the connection closes. Kept
//! byte-compatible with the Go implementation on purpose: the two run side by
//! side while the port is in progress, and the panel cannot tell them apart.

use serde::{Deserialize, Serialize};
use std::io::{self, BufRead, Read, Write};

/// The default unix socket. Mode 0660, owned root:opanel.
pub const SOCKET_PATH: &str = "/run/opanel/agent.sock";

/// Caps a single request or response. Actions move config text, not files;
/// anything larger is a bug or an attack.
pub const MAX_MESSAGE_BYTES: usize = 4 << 20;

/// One call.
#[derive(Debug, Deserialize)]
pub struct Request {
    /// Echoed back so a caller can correlate logs. No protocol meaning.
    #[serde(default)]
    pub id: String,
    /// Names the handler, e.g. "ping" or "systemd.restart".
    pub action: String,
    /// Pins the payload shape, so a stale panel fails loudly rather than
    /// being misread by a newer agent.
    #[serde(default)]
    pub version: i64,
    /// Action-specific input, decoded by the handler.
    #[serde(default)]
    pub payload: Option<serde_json::Value>,
}

/// The single reply.
#[derive(Debug, Serialize)]
pub struct Response {
    #[serde(skip_serializing_if = "String::is_empty")]
    pub id: String,
    pub ok: bool,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub error: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub code: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Value>,
}

impl Response {
    pub fn ok(id: &str, payload: serde_json::Value) -> Self {
        Self { id: id.to_owned(), ok: true, error: String::new(), code: String::new(), payload: Some(payload) }
    }

    pub fn failed(id: &str, code: Code, message: impl Into<String>) -> Self {
        Self {
            id: id.to_owned(),
            ok: false,
            error: message.into(),
            code: code.as_str().to_owned(),
            payload: None,
        }
    }
}

/// Returned in `Response.code` so callers branch without matching messages.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Code {
    UnknownAction,
    VersionMismatch,
    BadPayload,
    Denied,
    Internal,
    Timeout,
}

impl Code {
    pub fn as_str(self) -> &'static str {
        match self {
            Code::UnknownAction => "unknown_action",
            Code::VersionMismatch => "version_mismatch",
            Code::BadPayload => "bad_payload",
            Code::Denied => "denied",
            Code::Internal => "internal",
            Code::Timeout => "timeout",
        }
    }
}

/// Reads one newline-terminated request, refusing anything oversized before
/// it is allocated rather than after.
pub fn read_request<R: BufRead>(r: &mut R) -> io::Result<Request> {
    let mut line = Vec::new();
    // by_ref, because Read::take consumes the reader it is called on and the
    // caller still owns this one.
    r.by_ref()
        .take(MAX_MESSAGE_BYTES as u64 + 1)
        .read_until(b'\n', &mut line)?;
    if line.len() > MAX_MESSAGE_BYTES {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "request too large"));
    }
    if line.is_empty() {
        return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "empty request"));
    }
    serde_json::from_slice(&line).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
}

/// Writes one reply, newline-terminated.
pub fn write_response<W: Write>(w: &mut W, resp: &Response) -> io::Result<()> {
    let mut body = serde_json::to_vec(resp)?;
    body.push(b'\n');
    w.write_all(&body)?;
    w.flush()
}
